src/debug_tools/diff.rs

```diff
@@ -0,0 +1,507 @@
+//! Diffing tools to compare two tries against each other. Useful when you want
+//! to find where the tries diverge from one other.
+//!
+//! There are a few considerations when implementing the logic to create a trie
+//! diff:
+//! - What should be reported when the trie node structures diverge (eg. two
+//!   different node types proceeding a given common node)?
+//! - If there are multiple structural differences, how do we discover the
+//!   smallest difference to report back?
+//!
+//! If the node types between the tries (structure) are identical but some
+//! values are different, then these types of diffs are easy to detect and
+//! report the lowest difference. Structural differences are more challenging
+//! and a bit hard to report well. There are two approaches (only one currently
+//! is implemented) in how to detect structural differences:
+//! - Top-down search
+//! - Bottom-up search
+//!
+//! These two searches are somewhat self-explanatory:
+//! - Top-down will find the highest point of a structural divergence and report
+//!   it. If there are multiple divergences, then only the one that is the
+//!   highest in the trie will be reported.
+//! - Bottom-up (not implemented) is a lot more complex to implement, but will
+//!   attempt to find the smallest structural trie difference between the trie.
+//!   If there are multiple differences, then this will likely be what you want
+//!   to use.
+
+use std::fmt::{self, Debug};
+use std::{fmt::Display, ops::Deref};
+
+use ethereum_types::H256;
+
+use super::common::{
+    get_key_piece_from_node_no_branch_key, get_segment_from_node_and_key_piece, NodePath,
+};
+use crate::{
+    nibbles::Nibbles,
+    partial_trie::{HashedPartialTrie, Node, PartialTrie},
+    utils::TrieNodeType,
+};
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct TrieDiff {
+    pub latest_diff_res: Option<DiffPoint>,
+    // TODO: Later add a second pass for finding diffs from the bottom up (`earliest_diff_res`).
+}
+
+impl Display for TrieDiff {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(diff) = &self.latest_diff_res {
+            write!(f, "{}", diff)?;
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum DiffDetectionState {
+    NodeTypesDiffer, // Also implies that hashes differ.
+    HashDiffDetected,
+    NoDiffDetected,
+}
+
+impl DiffDetectionState {
+    fn pick_most_significant_state(&self, other: &Self) -> Self {
+        match self.get_int_repr() > other.get_int_repr() {
+            false => *other,
+            true => *self,
+        }
+    }
+
+    /// The integer representation also indicates the more "significant" state.
+    fn get_int_repr(&self) -> usize {
+        match self {
+            DiffDetectionState::NodeTypesDiffer => 2,
+            DiffDetectionState::HashDiffDetected => 1,
+            DiffDetectionState::NoDiffDetected => 0,
+        }
+    }
+}
```

Comment:
Nit: if we reorder the `Enum` definition as:

```rust
enum DiffDetectionState {
    NoDiffDetected = 0,
    HashDiffDetected,
    NodeTypesDiffer, // Also implies that hashes differ.
}
```

then we can directly access the discriminants by calling `as usize`, or if we want to have it explicit through this method, then it can just become (though a bit overkill IMO):

```rust
    fn get_int_repr(self) -> usize {
        self as usize
    }
```

Reply:
Nah I agree. I thought I tried this and it didn't work, but yeah casting directly is the nicest.